src/coreclr/jit/morph.cpp

```diff
@@ -13446,6 +13441,29 @@ GenTree* Compiler::fgOptimizeEqualityComparisonWithConst(GenTreeOp* cmp)
     {
         ssize_t op2Value = static_cast<ssize_t>(op2->IntegralValue());
 
+        // Optimizes (X & 1) == 1 to (X & 1)
+        //
+        //                        EQ/NE                   AND
+        //                        /  \                   /  \.
+        //                      AND   CNS 1  ->         x   CNS 1
+        //                     /   \.
+        //                    x   CNS 1
+        //
+        // The compiler requires jumps to have relop operands, so we do not fold that case.
+        if (op1->OperIs(GT_AND) && (op2Value == 1) && !(cmp->gtFlags & GTF_RELOP_JMP_USED))
+        {
+            if ((op1->gtGetOp1()->gtOper == GT_LCL_VAR) && op1->gtGetOp2()->IsIntegralConst(1) &&
```

Comment:
Why does restricting the `AND`'s operand to a local solve the regressions - what was their source?

Reply:
@SingleAccretion Well, I got pretty many regressions and It seemed suspicious to me that such a tiny optimization could cause this. I used jit-analyzer to figure out what was the actual diff between the dasms, but ended up just comparing both side by side and found out that the optimization introduced extra asm instructions (I'm not good at low-level engineering, I'm learning), so I assumed that the optimization had been applied in cased it should have been applied.
The restriction for `x` to be a local variable (This is how I understand the GT_LCL_VAR type) did the trick. I'm curious why @EgorBo said that `x` can be anything and would like to dig deeper to figure out what is the source of the regressions when `x` is anything. How do you debug such cases?

Reply:
> Well, I got pretty many regressions and It seemed suspicious to me that such a tiny optimization could cause this.

I agree, that is why it would be good to know; the restriction on the local may just be hiding it. Some regressions cannot be avoided (it is usual that, for example, we get some due to different register allocation), but we should have a good understanding of the causes to make decisions.

> How do you debug such cases?

I personally do it by: a) fully diffing with SPMI b) analyzing the cases individually by first diffing (as in `git diff`, or some online tool) the assembly, then the `JitDump`s (these ones are usually only `git diff`-able...), and debugging as necessary. I believe these days you can do this with the SPMI script directly by using the `-c` argument to pass the method context number that had the diffs.

Reply:
Thank you for the debugging advice. I've spent few days trying comparing dasm files. It is very hard for me to tell what is exactly wrong, but I found some suspicious facts:
1. Regressions mostly occur when 'x' is a pointer (or ref variable). 
Here is the diff example https://www.diffchecker.com/fQLJQ4YW
2. Regressions mostly occur in `MoveNext()` method 
<details>
  <summary>Example</summary>
 <img width="1201" alt="image" src="https://user-images.githubusercontent.com/18483133/147747869-a32338bd-d4fb-4ff3-89ad-2ea67fcc1b0f.png">
</details>
3. Some regressed dasms show terrible PerfScore. Here is the example https://www.diffchecker.com/BQUzlt9K

Reply:
Thank you for looking into this. I took a look myself, I agree it is a bit of a tricky case.

We lose tail duplication [because it only looks at relops for our purposes](https://github.com/dotnet/runtime/blob/56c4eec74595a5c38b8b8fd15124f7c0fca0bf7e/src/coreclr/jit/fgopt.cpp#L3167-L3200). Then because of that we lose relop forward substitution, but perhaps more importantly, the flow that RA can, apparently, better work with.

If we look at PerfScore instead of code size, things looks better (just a few small regressions), this holds both for `libraries.pmi` and `libraries_tests.pmi`. The PerfScore outliers you note are indeed curious. It appears for the "diff" method the altered flow causes the compiler to use synthetic scales (instead of profile data) for more blocks, and for more general loops to exist.

(Side note: `JitDisablePgo=1` doesn't appear to work well, so it is a bit hard to say why things are the way they are)

Overall, it appears doing this in lowering will be better after all.